tests/unit/src/test/scala/tests/DidFocusLspSuite.scala

```diff
@@ -122,10 +144,16 @@ class DidFocusLspSuite extends BaseLspSuite("did-focus") {
       didSaveA = server.didSave("a/src/main/scala/a/A.scala")(
         _.replace("Int", "String")
       )
+      // Wait until compilation against project a is started (before we invoke didFocus on project b)
+      _ <- Future { Thread.sleep(compileDelayMillis / 2) }
```

Comment:
I believe as an alternative we could also use
```
      _ <- server.waitFor(TimeUnit.SECONDS.toMillis(10)) // whatever time you want
```

Would just adding something like that in be enough so we don't have to pollute `Compilations` with `onStartCompilation`? It's sort of a bummer that we need to add that justs for tests.

Reply:
After seeing `TimeUnit.SECONDS.toMillis(10)` I wondered like ~5seconds if it's 10s transformed into millis or 10millis transformed to seconds :P
I suggest changing signature to `def waitFor(duration: FiniteDuration): Future[Unit] = Future { Thread.sleep(duration.toMillis) }`

Reply:
> Would just adding something like that be enough so we don't have to pollute Compilations with onStartCompilation? It's sort of a bummer that we need to add that just for tests.

Yeah, it's a bummer... but I don't think just adding a `waitFor` in the test fixes the problem.

### (1) Expected scenario
The diagram below is the expected test scenario:
- `didSave(A.scala)` start compilation
- and then `didSave(B.scala)` 
- and then compilation by `didSave(A.scala)` ends

```mermaid
sequenceDiagram
  participant Client
  participant Server

  Client->>Server: didSave(A.scala)
  rect rgb(191, 223, 255)
    Server->>Server: Start compile (by didSave)
    rect rgb(200, 150, 255)
      Client->>Server: didFocus
      Server->>Server: Compile by didFocus
    end
    Server->>Server: Finish compile (by didSave)
  end
```

### (2) didFocus too early

This (↓) is what we were:

- We invoke `didSave(A.scala)`
- but we call `didFocus(B.scala)` before the compilation (by didSave) starts
- then compilation by didSave(A.scala) starts and end.

```mermaid
sequenceDiagram
  participant Client
  participant Server

  Client->>Server: didSave(A.scala)
  rect rgb(200, 150, 255)
    Client->>Server: didFocus
    Server->>Server: Compile by didFocus
  end
  rect rgb(191, 223, 255)
    Server->>Server: Start compile (by didSave)

    Server->>Server: Finish compile (by didSave)
  end
```

We can solve this problem by adding a sleep before calling `didFocus` in the test.

### (3) didFocus too late

However, if we delay only before the `didFocus`, another issue occurs:

- didSave(A.scala)
- we sleep for a sec (before calling `didSave`)
- compilation by didSave starts and ends
- Then we call `didFocus`

```mermaid
sequenceDiagram
  participant Client
  participant Server

  Client->>Server: didSave(A.scala)
  rect rgb(191, 223, 255)
    Server->>Server: Start compile (by didSave)
    Server->>Server: Finish compile (by didSave)
  end
    rect rgb(200, 150, 255)
      Client->>Server: didFocus
      Server->>Server: Compile by didFocus
    end
```

In this scenario, the test passes :+1:. However, this is not what this test suite wants to test: according to the original PR https://github.com/scalameta/metals/pull/500, we try to test here: Metals triggers compilation by `didFocus` when **current compile** is going on (and it affects the focused buffer).
That's why we need a delay in compilation to make sure we call `didFocus` **during** ongoing compilation.

---

Maybe we can compromise not to test the "(1) expected scenario" and test (3). (Because I agree it's a bummer to add a hook in Compilation just for test).
On the other hand, maybe we can use this hook for other purposes in the future (not a test thing, I'm not sure... 🤔 )

Reply:
😆 thank you for the incredibly detailed examples. How do you like mermaid? Do you also have experience with plantuml? Any thoughts on the differences if you do?

Regarding the actual change, no worries. I think what you have is minimal enough it's not an issue.

Reply:
Yeah, I've been a big fan of plantuml, and quite new to mermaid :)
I couldn't find a big difference between plantuml and mermaid in terms of their functionalities, but I prefer mermaid rendered diagrams (just because their shape and color), and I found we can freely customize the styles by CSS and adding colored blocks by `rgb` is quite useful. 
On top of that,  being able to draw a diagram with "mermaid```...```" is pretty handy :+1:

> Regarding the actual change, no worries. I think what you have is minimal enough it's not an issue.

Thanks!